### PR TITLE
RavenDB-16940 Added operation to download or modify the database settings.

### PR DESCRIPTION
**Pull request by maciejaszyk**
https://issues.hibernatingrhinos.com/issue/RavenDB-16940

### CODE REVIEW

File: src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+namespace Raven.Client.ServerWide.Operations.Configuration
+{
+    public class DatabaseConfigurationSettings : IDynamicJson
+    {
+        public List<ConfigurationEntryValue> Settings { get; set; }
+
+        public DynamicJsonValue ToJson()
+        {
+            return new DynamicJsonValue {[nameof(Settings)] = new DynamicJsonArray(Settings.Select(x=>x.ToJson()))};
```
Comment (ml054) [inline src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs]: format

File: src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            return new DynamicJsonValue
+            {
+                [nameof(Metadata)] = Metadata.ToJson(),
+                [nameof(DatabaseValues)] = DynamicJsonValue.Convert(DatabaseValues),
+                [nameof(ServerValues)] = DynamicJsonValue.Convert(ServerValues)
+            };
+        }
+    }
+
+    public class ConfigurationEntryMetadata : IDynamicJson
```
Comment (ml054) [inline src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs]: don't we already have that DTO somewhere?

File: src/Raven.Client/ServerWide/Operations/Configuration/PutDatabaseConfigurationSettingsOperation.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            private readonly string _databaseName;
+
+            public PutDatabaseConfigurationSettingsCommand(JsonOperationContext context, Dictionary<string, string> configurationSettings, string databaseName)
+            {
+                if (context is null)
+                    throw new ArgumentNullException(nameof(context));
+                if (configurationSettings is null)
+                    throw new ArgumentNullException(nameof(configurationSettings));
+                _databaseName = databaseName ?? throw new ArgumentNullException(nameof(databaseName));
+                _configurationSettings = DocumentConventions.Default.Serialization.DefaultConverter.ToBlittable(configurationSettings, context);
```
Comment (ml054) [inline src/Raven.Client/ServerWide/Operations/Configuration/PutDatabaseConfigurationSettingsOperation.cs:41]: are you sure we have to use here default document convensions?

File: src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs [hunk 4]
```diff
@@ -72,6 +73,40 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
             }
         }
 
+        //TODO
+        [RavenAction("/databases/*/admin/configuration/settings2", "GET", AuthorizationStatus.DatabaseAdmin, IsDebugInformationEndpoint = true)]
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs]: let's change that to `/databases/*/admin/record` and reuse code from `AdminDatabasesHandler`

File: src/Raven.Client/ServerWide/Operations/Configuration/PutDatabaseConfigurationSettingsOperation.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Http;
+using Raven.Client.Json;
+using Raven.Client.Util;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
+{
+    class PutDatabaseConfigurationSettingsOperation : IServerOperation
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/PutDatabaseConfigurationSettingsOperation.cs]: I think it should be `IMaintenanceOperation`
Also it needs to be public

File: src/Raven.Client/ServerWide/Operations/Configuration/GetDatabaseConfigurationSettingsOperation.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
+using System.Net.Http;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Http;
+using Raven.Client.Json.Serialization;
+using Sparrow.Json;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
+{
+    public class GetDatabaseConfigurationSettingsOperation : IServerOperation<DatabaseConfigurationSettings>
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/GetDatabaseConfigurationSettingsOperation.cs]: IMaintenanceOperation

File: src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Sparrow;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Client.ServerWide.Operations.Configuration
+{
+    public class DatabaseConfigurationSettings : IDynamicJson
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs]: DatabaseSettings

File: src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public Dictionary<string, string> Values { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/DatabaseConfigurationSettings.cs]: Settings

File: src/Raven.Client/ServerWide/Operations/Configuration/GetDatabaseConfigurationSettingsOperation.cs (hunk 6 above)
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/GetDatabaseConfigurationSettingsOperation.cs]: GetDatabaseSettingsOperation

File: src/Raven.Client/ServerWide/Operations/Configuration/PutDatabaseConfigurationSettingsOperation.cs (hunk 5 above)
Comment (ppekrol) [inline src/Raven.Client/ServerWide/Operations/Configuration/PutDatabaseConfigurationSettingsOperation.cs]: PutDatabaseSettingsOperation

**Review by arekpalinski — APPROVED** [review]

File: src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs [hunk 9]
```diff
@@ -72,6 +75,12 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
             }
         }
 
+        [RavenAction("/databases/*/admin/record", "GET", AuthorizationStatus.DatabaseAdmin)]
+        public async Task GetSettingForClient()
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs]: rename to `GetDatabaseRecord`

File: src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs [hunk 10]
```diff
@@ -72,6 +75,12 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
...
+        {
+            await SendDatabaseRecord(Database.Name, ServerStore, HttpContext, ResponseBodyStream());
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs:81]: Sanity check:
@ayende The `/admin/databases` endpoint is only accessible by operator. Now we are exposing to DatabaseAdmin record of the database that they can manage. Which sounds reasonable, do you have any objections? Is there anything in the record that we would not want to expose to DatabaseAdmin? I cannot think of anything, but would like an additional person to confirm.

Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs:81]: DatabaseAdmin should be able to see everything there, yes.

Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs:81]: but what is the scenario where you have only db admin right and need it?

Comment (ppekrol) [reply src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs:81]: We are using this to download settings, but I didn't want to have X-number of endpoints for all of the properties, so we decided to expose whole record and just take the value of the property in the client
e.g. in this case admin can get settings and then change them

Comment (ayende) [reply src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs:81]: Okay, make sense.

File: src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs (hunk 9 above)
Comment (maciejaszyk) [reply src/Raven.Server/Documents/Handlers/Admin/AdminConfigurationHandler.cs]: done

**Pull request merged**